### PR TITLE
Unify creation and retry connection for portal and relay

### PR DESCRIPTION
**Pull request by davide-baldo**
Implements https://github.com/build-trust/ockam/issues/4789

The Medic callback was refactored into a more usable trait called `Replacer`. Its responsibility is to spawn and keep track of resources of relays and portals (or possible other elements in the future).

The differentiation between "first connection code" and later attempts was removed, less repetitive code and more tested. This should result in more robust implementations.
Both relays and portals didn't handle the concept of connection being down, every field was always returned even if it was old/stale data of a closed connection. This was addressed, and the answers now have optional fields when the information is not available due to the connection being down.
In portal creation there is an extra flag called `--no-validation` was added. This flag skips the synchronous validation, and allows the operator to create an inlet even if the route is not fully reachable.

### CODE REVIEW

File: implementations/rust/ockam/ockam_command/src/relay/create.rs [hunk 1]
```diff
@@ -50,6 +49,10 @@ pub struct CreateCommand {
     /// Authorized identity for secure channel connection
     #[arg(long, id = "AUTHORIZED")]
     pub authorized: Option<Identifier>,
+
+    /// Relay address to use. By default, inherits the relay name.
+    #[arg(long)]
+    relay_address: Option<String>,
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_command/src/relay/create.rs:55]: In what scenario is this useful? IIUC, this is to define the name of the relay that will be created while the alias is only used in the node to store it in the registry?

**Review by adrianbenavides — COMMENTED** [review]
I've done a first pass over the `ockam_command` crate and a little around how the inlets are created. There will be some conflicts with sanjo's policy PR, but they should be easy to fix.

File: implementations/rust/ockam/ockam_api/src/session/sessions.rs [hunk 2]
```diff
@@ -79,57 +154,86 @@ impl Session {
     ///
     /// # Arguments
     ///
-    /// * `ping_route` - The route to send pings to
     /// * `key` - The key to identify the session, usually adding the kind of the session
     ///           with the key used within the service registry is the way to gos
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/session/sessions.rs]: This is outdated now right?

File: implementations/rust/ockam/ockam_api/src/session/sessions.rs [hunk 3]
```diff
@@ -79,57 +154,86 @@ impl Session {
...
-    pub fn new(ping_route: Route, key: String) -> Self {
+    pub fn new(replace: impl SessionReplacer) -> Self {
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/session/sessions.rs]: Typo? `replace` -> `replacer`

File: implementations/rust/ockam/ockam_api/src/nodes/service/relay.rs [hunk 4]
```diff
@@ -122,81 +122,89 @@ impl NodeManager {
...
+                    }
+                })?;
+
+        if self.registry.relays.contains_key(&alias).await {
+            let message = format!("A relay with the name '{alias}' already exists");
+            return Err(ockam_core::Error::new(
+                Origin::Node,
+                Kind::AlreadyExists,
+                message,
+            ));
```
Comment (adrianbenavides) [inline implementations/rust/ockam/ockam_api/src/nodes/service/relay.rs]: Why is the relay created before checking if it already exists? In the `create_inlet` function it's doing the check first.

File: implementations/rust/ockam/ockam_command/src/relay/create.rs (hunk 1 above)
Comment (davide-baldo) [reply implementations/rust/ockam/ockam_command/src/relay/create.rs:55]: By default the alias defines both the name of the relay and the address of the relay.
Normally, you would only use alias, but if you want to use the same address for multiple projects relaying toward the same node, then you need relay_address.
Another simple use case is to keep compatibility with a previously named address while renaming the relay itself.

File: implementations/rust/ockam/ockam_api/src/session/sessions.rs (hunk 2 above)
Comment (davide-baldo) [reply implementations/rust/ockam/ockam_api/src/session/sessions.rs]: yes! deleting...

File: implementations/rust/ockam/ockam_api/src/nodes/service/relay.rs (hunk 4 above)
Comment (davide-baldo) [reply implementations/rust/ockam/ockam_api/src/nodes/service/relay.rs]: no particular reason, it makes sense to delay the creation after full validation

**Review by adrianbenavides — APPROVED** [review]

**Pull request merged**